### PR TITLE
docs(BulkWrite): adds documentation to BulkWrite

### PR DESCRIPTION
**Pull request by daprahamian**
## Description

Adds documentation for multiple methods and properties
associated with bulk writes. Also marks certain methods as
internal.

Fixes NODE-1800
Fixes NODE-1938

**What changed?**

Added more documentation to the BulkOperationBase. Includes, but is not limited to:
- Updating language
- Adding examples for how to add various operations to the bulk op
- Correcting the jsdoc for getters such that the description and type actually show up
- Marking certain methods as internal by adding an `@ignore` tag

**Are there any files to ignore?**
- `lib/bulk/ordered.js` and `lib/bulk/unordered.js` have trivial changes.

### CODE REVIEW

File: lib/bulk/common.js [hunk 1]
```diff
@@ -55,60 +55,69 @@ class Batch {
...
+   *
+   * @class
+   */
   constructor(bulkResult) {
     this.result = bulkResult;
   }
 
   /**
-   * @return {boolean} ok Did bulk operation correctly execute
+   * Is true if the bulk operation correctly execute
```
Comment (kvwalker) [inline lib/bulk/common.js]: Might be worded better as `Returns true if the bulk operation correctly executes`

File: lib/bulk/common.js [hunk 2]
```diff
@@ -55,60 +55,69 @@ class Batch {
...
- * @class
- * @return {BulkWriteResult} a BulkWriteResult instance
+ * @classdesc
+ * The result of a bulk write.
  */
 class BulkWriteResult {
+  /**
+   * Create a new BulkWriteResult instance (INTERNAL TYPE, do not instantiate directly)
+   *
+   * @class
```
Comment (mbroadst) [inline lib/bulk/common.js]: I believe `@class` is superfluous if you are using ES6 classes

File: lib/bulk/common.js [hunk 3]
```diff
@@ -220,7 +229,7 @@ class BulkWriteResult {
   }
 
   /**
-   * @return {BulkWriteResult} a BulkWriteResult instance
+   * @return {object}
```
Comment (mbroadst) [inline lib/bulk/common.js]: No docs here?

File: lib/bulk/common.js [hunk 4]
```diff
@@ -242,25 +251,30 @@ class BulkWriteResult {
...
- * @class
- * @return {WriteConcernError} a WriteConcernError instance
+ * @classdesc An error representing a failure by the server to apply the requested write concern to the bulk operation.
  */
 class WriteConcernError {
+  /**
+   * Create a new WriteConcernError instance (INTERNAL TYPE, do not instantiate directly)
+   *
+   * @class
+   * @return {WriteConcernError} a WriteConcernError instance
```
Comment (mbroadst) [inline lib/bulk/common.js]: we don't document return values for most constructors, should we be doing that here?

File: lib/bulk/common.js [hunk 5]
```diff
@@ -242,25 +251,30 @@ class BulkWriteResult {
...
 /**
- * Create a new WriteConcernError instance (INTERNAL TYPE, do not instantiate directly)
- *
- * @class
- * @return {WriteConcernError} a WriteConcernError instance
+ * @classdesc An error representing a failure by the server to apply the requested write concern to the bulk operation.
  */
 class WriteConcernError {
+  /**
+   * Create a new WriteConcernError instance (INTERNAL TYPE, do not instantiate directly)
```
Comment (mbroadst) [inline lib/bulk/common.js]: nit: we have traditionally called out notes in bold, like:
```
Create a new WriteConcernError

**NOTE:** Internal type, do not instantiate directly
```
I don't know if there's another jsdoc tag for this

File: lib/bulk/common.js [hunk 6]
```diff
@@ -531,15 +551,17 @@ function handleMongoWriteConcernError(batch, bulkResult, err, callback) {
...
- * @param {BulkWriteResult} result The result of the bulk write operation
- * @return {BulkWriteError} A BulkWriteError instance
- * @extends {MongoError}
+ * @classdesc An error indicating an unsuccessful Bulk Write
  */
 class BulkWriteError extends MongoError {
+  /**
+   * Creates a new BulkWriteError
+   *
+   * @class
```
Comment (mbroadst) [inline lib/bulk/common.js]: no need for `@class`

File: lib/bulk/common.js [hunk 7]
```diff
@@ -531,15 +551,17 @@ function handleMongoWriteConcernError(batch, bulkResult, err, callback) {
...
 
 /**
- * Creates a new BulkWriteError
- *
- * @class
- * @param {Error|string|object} message The error message
- * @param {BulkWriteResult} result The result of the bulk write operation
- * @return {BulkWriteError} A BulkWriteError instance
- * @extends {MongoError}
+ * @classdesc An error indicating an unsuccessful Bulk Write
```
Comment (mbroadst) [inline lib/bulk/common.js]: I also think `@classdesc` is superfluous

File: lib/bulk/common.js [hunk 8]
```diff
@@ -552,24 +574,25 @@ class BulkWriteError extends MongoError {
...
+   * Creates a new FindOperators object. For internal use only. Do not invoke directly.
    * @param {OrderedBulkOperation|UnorderedBulkOperation} bulkOperation
    */
   constructor(bulkOperation) {
     this.s = bulkOperation.s;
   }
 
   /**
-   * Add a single update document to the bulk operation
+   * Add a multiple update operation to the bulk operation
```
Comment (mbroadst) [inline lib/bulk/common.js]: why not just "Adds an update operation to the bulk operation"?

File: lib/bulk/common.js [hunk 9]
```diff
@@ -55,60 +55,69 @@ class Batch {
...
+   *
+   * **NOTE:** Internal Type, do not instantiate directly
+   */
   constructor(bulkResult) {
     this.result = bulkResult;
   }
 
   /**
-   * @return {boolean} ok Did bulk operation correctly execute
+   * Evaluates to true if the bulk operation correctly execute
```
Comment (kvwalker) [inline lib/bulk/common.js]: `execute` should be `executes` (or `executed`?)

File: lib/bulk/common.js [hunk 10]
```diff
@@ -282,39 +295,44 @@ class WriteConcernError {
...
+   * Write error original bulk operation index.
+   * @type {number}
    */
   get index() {
     return this.err.index;
   }
 
   /**
-   * @return {string} errmsg Write concern error message.
+   * Write  error message.
```
Comment (kvwalker) [inline lib/bulk/common.js]: I think there's an extra space here between Write and error

**Comment by daprahamian** [comment]
@dandv I have corrected the class description to reference `delete(One)?` instead of the `remove` functions. I've also added an example for `replaceOne`, and added an `updateOne` example using a pipeline to leverage previous queried values (only available in 4.2).

File: lib/bulk/common.js (hunk 3 above)
Comment (mbroadst) [reply lib/bulk/common.js]: SGTM

**Pull request merged**